### PR TITLE
[react] Enable Awaited<ReactNode> in experimental release channels

### PR DESCRIPTION
**Pull request by eps1lon**
Fixes https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/68909

`type T = string | Promise<T>` is rejected for a reason. The previous trick by using interface just masked the problem until you `Awaited<T>` and got `TS2589: Type instantiation is excessively deep and possibly infinite.`

### CODE REVIEW

File: types/react/experimental.d.ts [hunk 1]
```diff
@@ -42,10 +42,22 @@ declare const UNDEFINED_VOID_ONLY: unique symbol;
...
 
 declare module "." {
-    // Need an interface to not cause ReactNode to be a self-referential type.
-    interface PromiseLikeOfReactNode extends PromiseLike<ReactNode> {}
+    /**
+     * @internal Use `Awaited<ReactNode>` instead
+     */
+    // Helper type to enable `Awaited<ReactNode>`.
+    // Must be a copy of the non-thenables of `ReactNode`.
+    type AwaitedReactNode =
```
Comment (eps1lon) [inline types/react/experimental.d.ts:50]: A bit annoying that we have to introduce this type. TypeScript should just always flatten `Promise<Promise<T>>` to `Promise<T>` and allow `type T = U | Promise<T>`. Neither `await` nor `then()` will observe the recursive Promise.

Comment (weswigham) [reply types/react/experimental.d.ts:50]: @rbuckton had a PR that introduced that behavior at one point, iirc, but it had some [issues in the real world](https://github.com/microsoft/TypeScript/pull/37610). The `Awaited` type as it exists today is a bit of a compromise in that regard.

**Comment by jtbandes** [comment]
Hi there, I'm running into some issues where the addition of some "thenable" types in `@types/react` is triggering the [@typescript-eslint/promise-function-async](https://typescript-eslint.io/rules/promise-function-async/) lint rule. For example:

```ts

```

**Comment by jtbandes** [comment]
Hi there, I'm running into some issues where the addition of some "thenable" types in `@types/react` is triggering the [@typescript-eslint/promise-function-async](https://typescript-eslint.io/rules/promise-function-async/) lint rule. For example:

```ts
...
{
  wrapper: class Wrapper extends React.Component<React.PropsWithChildren> {
    public override componentDidCatch(err: Error) {
      error = err;
    }
    // ↓ error: Functions that return promises must be async.
    public override render() {
      return this.props.children;
    }
  },
},
```

It seems that TypeScript and typescript-eslint are noticing that `props.children`, which is typed as `ReactNode`, can now contain a Promise in some situations and therefore triggering this error.

In some cases I've been able to work around these by replacing `ReactNode` types with `JSX.Element`. But this doesn't work for the example above where the type of `children` is out of my control (it comes from `React.PropsWithChildren`).

Is this expected behavior?

**Comment by eps1lon** [comment]
@jtbandes I don't know what React could do here. Usercode will never handle the return value of `render` so the rationale of the rule doesn't apply. Can you file an issue with eslint-plugin-typescript and tag me so that we can discuss how to move forward?

**Comment by jtbandes** [comment]
That's fair, I will file an issue upstream. However in my [attempt to reproduce in a playground](https://typescript-eslint.io/play/#ts=5.4.5&fileType=.tsx&code=JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdAbhMACZy4A2a6AYhBDhYAHjCwA7Vumx4YAOgDCJSOIkwAPDPxyACsTCoA6sBgALBaeCdWlcQD44Abzpw4YAK4AjTsFxwIjFhQUGxYcLasQQAUAJROLq7hWDDuUOJwZsCocmD62biW1rb0rgC%2BdOVAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6AewFtLlFaAM1hMy%2BSpya0AhskKj0URNC7RI4MAF8QmoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false), I did not encounter the error. I suspect this is because the "experimental" types may not be getting pulled in, so it doesn't think the "thenable" type is present. What would cause them to be used by TypeScript?

**Comment by eps1lon** [comment]
> What would cause them to be used by TypeScript?

`import 'react/experimental'` or `/// <reference types="react/experimental" />`

**Comment by jtbandes** [comment]
The TS playground is interesting because I didn’t include the experimental types intentionally, yet it accepts a promise as assignable to ReactNode. This is the same in my actual project — I haven’t included the experimental types yet they seem to be included somehow…

**Comment by jtbandes** [comment]
☹️ 
```
$ grep -R 'react/experimental' .
...
./node_modules/next/types/index.d.ts:/// <reference types="react/experimental" />
```

**Comment by jtbandes** [comment]
Filed https://github.com/typescript-eslint/typescript-eslint/issues/9341

**Comment by jtbandes** [comment]
For anyone discovering this issue, the solution identified in https://github.com/typescript-eslint/typescript-eslint/issues/9341 is to annotate the return type explicitly:

```ts
class extends React.Component<React.PropsWithChildren> {
  public override render(): React.ReactNode {  // no error
    return this.props.children;
  }
}
```

**Pull request merged**